tfsdk/block_test.go

```diff
@@ -5,13 +5,765 @@ import (
 	"testing"
 
 	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
 	"github.com/hashicorp/terraform-plugin-framework/diag"
 	testtypes "github.com/hashicorp/terraform-plugin-framework/internal/testing/types"
 	"github.com/hashicorp/terraform-plugin-framework/types"
 	"github.com/hashicorp/terraform-plugin-go/tfprotov6"
 	"github.com/hashicorp/terraform-plugin-go/tftypes"
 )
 
+func TestBlockModifyPlan(t *testing.T) {
+	t.Parallel()
+
+	schema := func(blockPlanModifiers AttributePlanModifiers, nestedAttrPlanModifiers AttributePlanModifiers) Schema {
+		return Schema{
+			Blocks: map[string]Block{
+				"test": {
+					Attributes: map[string]Attribute{
+						"nested_attr": {
+							Type:          types.StringType,
+							Required:      true,
+							PlanModifiers: nestedAttrPlanModifiers,
+						},
+					},
+					NestingMode:   BlockNestingModeList,
+					PlanModifiers: blockPlanModifiers,
+				},
+			},
+		}
+	}
+
+	schemaTfValue := func(nestedAttrValue string) tftypes.Value {
+		return tftypes.NewValue(
+			tftypes.Object{
+				AttributeTypes: map[string]tftypes.Type{
+					"test": tftypes.List{
+						ElementType: tftypes.Object{
+							AttributeTypes: map[string]tftypes.Type{
+								"nested_attr": tftypes.String,
+							},
+						},
+					},
+				},
+			},
+			map[string]tftypes.Value{
+				"test": tftypes.NewValue(
+					tftypes.List{
+						ElementType: tftypes.Object{
+							AttributeTypes: map[string]tftypes.Type{
+								"nested_attr": tftypes.String,
+							},
+						},
+					},
+					[]tftypes.Value{
+						tftypes.NewValue(
+							tftypes.Object{
+								AttributeTypes: map[string]tftypes.Type{
+									"nested_attr": tftypes.String,
+								},
+							},
+							map[string]tftypes.Value{
+								"nested_attr": tftypes.NewValue(tftypes.String, nestedAttrValue),
+							},
+						),
+					},
+				),
+			},
+		)
+	}
+
+	var schemaNullTfValue tftypes.Value = tftypes.NewValue(
+		tftypes.Object{
+			AttributeTypes: map[string]tftypes.Type{
+				"test": tftypes.List{
+					ElementType: tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"nested_attr": tftypes.String,
+						},
+					},
+				},
+			},
+		},
+		map[string]tftypes.Value{
+			"test": tftypes.NewValue(
+				tftypes.List{
+					ElementType: tftypes.Object{
+						AttributeTypes: map[string]tftypes.Type{
+							"nested_attr": tftypes.String,
+						},
+					},
+				},
+				nil,
+			),
+		},
+	)
+
+	modifyAttributePlanRequest := func(attrPath *tftypes.AttributePath, schema Schema, configValue, planValue, stateValue string) ModifyAttributePlanRequest {
```

Comment:
not a big deal, but if we make config, plan, and state members of a struct  type (e.g.

```go
type values struct {
config string
plan string
state string
}
```
) it may be a little easier to follow inside tests what the values are getting set to, because we can use what are essentially named arguments. I keep finding myself scrolling back up here to see what order they're in (especially since my brain keeps thinking it's state, config, plan).

Reply:
I agree! It is less than ideal to not have that field identifier context for knowing which value is for which data. I like this approach and will followup with the test refactoring.